File: PerfectSquares/PerfectSquares.py
class Solution(object):
    _dp = [0]
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        dp = self._dp
        while len(dp) <= n:
            sol = 10000
            m = int(len(dp)**0.5 + 1)
            for i in range(1, m):
                sol = min(sol, dp[-i*i])
            dp += sol + 1,
        return dp[n]

File: PerfectSquares/PerfectSquares.cpp
class Solution {
public:
    int numSquares(int n) {
        if(n <= 0)
        	return 0;
        static vector<int> dp({0});
        while(dp.size() <= n) {
        	int m = dp.size();
        	int cnt = INT_MAX;
        	for(int i = 1; i * i <= m; i++) {
        		cnt = min(cnt, dp[m - i * i] + 1);
        	}
        	dp.push_back(cnt);
        }
        return dp[n];
    }
};